orangecontrib/text/widgets/owtwitter.py

```diff
@@ -0,0 +1,275 @@
+            self.set_key_list()
+
+    def accept(self):
+        if self.key not in self.recent_keys:
+            self.check_api_key()
+        self.parent.update_api(self.key)
+        super().accept()
+
+    def key_selected(self, n):
+        key = self.recent_keys[n]
+        self.recent_keys.pop(n)
+        self.recent_keys.insert(0, key)
+        self.set_key_list()
+
+    def secret_changed(self):
+        pass
+
+
+class Output:
+    CORPUS = "Corpus"
+
+
+def api_required(func):
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if not self.api:
+            self.open_set_api_key_dialog_button.setFocus()
+            tip = 'Please provide a valid API key in order to get the data.'
+            QtGui.QToolTip.showText(QtGui.QCursor.pos(), tip)
+        else:
+            return func(self, *args, **kwargs)
+    return wrapper
+
+
+def word_list_required(func):
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if not self.query_words:
+            self.query_words_text_edit.setFocus()
+            tip = 'Please input query words.'
+            QtGui.QToolTip.showText(QtGui.QCursor.pos(), tip)
+        else:
+            return func(self, *args, **kwargs)
+    return wrapper
+
+
+class OWTwitter(OWWidget):
+    """
+    Attributes:
+        key (twitter.Credentials): Twitter API key/secret holder
+        api (twitter.TwitterAPI): Twitter API obj
+    """
+    name = "Twitter"
+    description = "Loads tweets from the Twitter API."
+    icon = "icons/Twitter.svg"
+
+    outputs = [(Output.CORPUS, Corpus)]
+    want_main_area = False
+    resizing_enabled = False
+
+    recent_api_keys = Setting([])
+    query_words = Setting('')
+    max_tweets = Setting(100)
+    language = Setting('')
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.key = self.recent_api_keys[0] if self.recent_api_keys else None
+        self.api = None
+
+        # Root box.
+        parameter_box = gui.widgetBox(self.controlArea, addSpace=True)
+
+        # API key box.
+        api_key_box = gui.widgetBox(parameter_box, orientation=0)
+        # Valid API key feedback.
+        self.api_key_valid_label = gui.label(api_key_box, self, "")
+        self.update_validity_icon()
+        self.api_key_valid_label.setMaximumSize(self.api_key_valid_label.sizeHint())
+        # Set API key button.
+        self.open_set_api_key_dialog_button = gui.button(api_key_box, self, 'Article API key',
+                                                         callback=self.open_set_api_key_dialog,
+                                                         tooltip="Set the API key for this widget.")
+
+        self.open_set_api_key_dialog_button.setFocusPolicy(QtCore.Qt.NoFocus)
+
+        # Queries configuration
+        layout = QtGui.QFormLayout()
+
+        # Max tweets
+        self.max_tweets_spin = gui.spin(self, self, 'max_tweets', minv=1, maxv=1000)
```

Comment:
👍 I suggest we are not more restrictive than tweepy — if they allow more we should also. I'm sure @kernc would agree 😉